### ISSUE TITLE
Getting panic: inconsistent label cardinality when starting

### ISSUE THREAD

**Issue by jeuniii** [wrouesnel/postgres_exporter#98]
I followed the supplied commands to `make` the project . While connecting to  my local Postgres Db,

I am getting : 

```
export DATA_SOURCE_NAME="postgresql://postgres:postgres@localhost:5432/postgres?sslmode=disable"

# ./postgres_exporter
INFO[0000] Established new database connection.          source=postgres_exporter.go:951
INFO[0000] Semantic Version Changed: 0.0.0 -> 9.4.1      source=postgres_exporter.go:904
panic: inconsistent label cardinality

goroutine 17 [running]:
postgres_exporter-master/vendor/github.com/prometheus/client_golang/prometheus.MustNewConstMetric(0xc42021c770, 0x3, 0x0, 0xc42025b120, 0x2, 0x2, 0x0, 0x0)
	/usr/local/go/src/postgres_exporter-master/vendor/github.com/prometheus/client_golang/prometheus/value.go:172 +0xb3
main.queryNamespaceMapping(0xc420184000, 0xc42014e280, 0x7c9e3f, 0x10, 0xc420129180, 0x2, 0x2, 0xc420139500, 0xc420139b90, 0x0, ...)
	/usr/local/go/src/postgres_exporter-master/postgres_exporter.go:853 +0x1200
main.queryNamespaceMappings(0xc420184000, 0xc42014e280, 0xc4201391a0, 0xc420139b90, 0x0)
	/usr/local/go/src/postgres_exporter-master/postgres_exporter.go:868 +0x321
main.(*Exporter).scrape(0xc420182000, 0xc420184000)
	/usr/local/go/src/postgres_exporter-master/postgres_exporter.go:996 +0x58d
main.(*Exporter).Collect(0xc420182000, 0xc420184000)
	/usr/local/go/src/postgres_exporter-master/postgres_exporter.go:747 +0x39
main.(*Exporter).Describe(0xc420182000, 0xc420144120)
	/usr/local/go/src/postgres_exporter-master/postgres_exporter.go:740 +0xb0
postgres_exporter-master/vendor/github.com/prometheus/client_golang/prometheus.(*registry).Register.func1(0x97ab20, 0xc420182000, 0xc420144120)
	/usr/local/go/src/postgres_exporter-master/vendor/github.com/prometheus/client_golang/prometheus/registry.go:218 +0x3b
created by postgres_exporter-master/vendor/github.com/prometheus/client_golang/prometheus.(*registry).Register
	/usr/local/go/src/postgres_exporter-master/vendor/github.com/prometheus/client_golang/prometheus/registry.go:217 +0xbc
```
I am unsure about why the `panic: inconsistent label cardinality` error occurs.

**Comment by wrouesnel** [comment]
Could you run with `-log.level=debug` so we can determine which exporter section is triggering it?

**Comment by jeuniii** [comment]
Thank you for your reply. Here is the info.

```
$ ./postgres_exporter -log.level=debug
INFO[0000] Established new database connection.          source=postgres_exporter.go:951
DEBU[0000] Querying Postgres Version                     source=postgres_exporter.go:887
INFO[0000] Semantic Version Changed: 0.0.0 -> 9.4.1      source=postgres_exporter.go:904
DEBU[0000] procpid is being forced to discard due to version incompatibility.  source=postgres_exporter.go:496
DEBU[0000] Querying pg_setting view                      source=pg_setting.go:17
DEBU[0000] Querying namespace:  pg_stat_database_conflicts  source=postgres_exporter.go:867
DEBU[0000] Querying namespace:  pg_locks                 source=postgres_exporter.go:867
DEBU[0000] Querying namespace:  pg_stat_replication      source=postgres_exporter.go:867
DEBU[0000] Querying namespace:  pg_stat_activity         source=postgres_exporter.go:867
DEBU[0000] Querying namespace:  pg_stat_bgwriter         source=postgres_exporter.go:867
DEBU[0000] Querying namespace:  pg_stat_database         source=postgres_exporter.go:867
DEBU[0000] blks_icache_hit [1 template1]                 source=postgres_exporter.go:852
panic: inconsistent label cardinality

goroutine 17 [running]:
postgres_exporter-master/vendor/github.com/prometheus/client_golang/prometheus.MustNewConstMetric(0xc42020ef50, 0x3, 0x0, 0xc42025b3a0, 0x2, 0x2, 0x0, 0x0)
	/usr/local/go/src/postgres_exporter-master/vendor/github.com/prometheus/client_golang/prometheus/value.go:172 +0xb3
main.queryNamespaceMapping(0xc420186000, 0xc420150280, 0x7c9e3f, 0x10, 0xc420146fe0, 0x2, 0x2, 0xc420139470, 0xc420139bc0, 0x0, ...)
	/usr/local/go/src/postgres_exporter-master/postgres_exporter.go:853 +0x1200
main.queryNamespaceMappings(0xc420186000, 0xc420150280, 0xc4201391d0, 0xc420139bc0, 0x0)
	/usr/local/go/src/postgres_exporter-master/postgres_exporter.go:868 +0x321
main.(*Exporter).scrape(0xc420184000, 0xc420186000)
	/usr/local/go/src/postgres_exporter-master/postgres_exporter.go:996 +0x58d
main.(*Exporter).Collect(0xc420184000, 0xc420186000)
	/usr/local/go/src/postgres_exporter-master/postgres_exporter.go:747 +0x39
main.(*Exporter).Describe(0xc420184000, 0xc420144120)
	/usr/local/go/src/postgres_exporter-master/postgres_exporter.go:740 +0xb0
postgres_exporter-master/vendor/github.com/prometheus/client_golang/prometheus.(*registry).Register.func1(0x97ab20, 0xc420184000, 0xc420144120)
	/usr/local/go/src/postgres_exporter-master/vendor/github.com/prometheus/client_golang/prometheus/registry.go:218 +0x3b
created by postgres_exporter-master/vendor/github.com/prometheus/client_golang/prometheus.(*registry).Register
	/usr/local/go/src/postgres_exporter-master/vendor/github.com/prometheus/client_golang/prometheus/registry.go:217 +0xbc
```

**Comment by wrouesnel** [comment]
Looks like you've found a case via your config where something I've assumed was fixed wasn't. I'll need to figure out a test case since the integration tests do test against postgres 9.4

**Comment by jeuniii** [comment]
I see. For me this is the first time Im setting up postgres_exporter.

Further details about my environment are : 

```
$ cat /etc/redhat-release 
Red Hat Enterprise Linux Server release 7.3 (Maipo)

$ go version
go version go1.9 linux/amd64
```

Builiding the project was also successful.

```
# make
make -C tools
make[1]: Entering directory `/usr/local/go/src/postgres_exporter-master-BK/tools'
make[1]: Nothing to be done for `DEFAULT'.
make[1]: Leaving directory `/usr/local/go/src/postgres_exporter-master-BK/tools'
gometalinter --disable-all --enable=gofmt --vendor
Using 8 processes
gometalinter -j 8 --deadline=30s --disable=gotype --disable=gocyclo 
for pkg in postgres_exporter-master-BK ; do \
	go test -v -covermode count -coverprofile=.coverage/$(echo $pkg | tr '/' '-').out $pkg ; \
done
=== RUN   Test
OK: 3 passed
--- PASS: Test (0.00s)
PASS
coverage: 11.6% of statements
ok  	postgres_exporter-master-BK	0.007s
gocovmerge .coverage/postgres_exporter-master-BK.out > cover.test.out
```
Let me know if you need anymore info.

**Comment by wrouesnel** [comment]
I have a feeling this issue occurs when parsing of metric values fails for an entry. The solution here is probably the ugly one of simply discarding the old description and rebuilding it when this happens, since we're not trying to enforce convention - just pass through what PG adds.

I still need to cook up a test case though to add some bounds on what the behaviour should be.

**Comment by keithf4** [comment]
I'm seeing this same error when I try and use the following query for iostats. Running the query straight in postgres works fine.
```
ccp_statio_user_objects:
  query: "SELECT TS.spcname AS tbl_space,
              schemaname,
              statio.relname AS tablename,
              statio.indexrelname AS indexname,
              pg_relation_size(indexrelid) AS size,
              idx_blks_read AS rel_blks_read,
              idx_blks_hit AS rel_blks_hit,
              idx_blks_hit+idx_blks_read AS rel_blks,
              CASE WHEN (idx_blks_hit+idx_blks_read)>0 THEN round(100*idx_blks_hit/(idx_blks_hit+idx_blks_read)) ELSE 0 END AS hit_pct
            FROM pg_statio_user_indexes statio
              JOIN pg_class C ON (C.oid = indexrelid)
              LEFT JOIN
                pg_tablespace TS ON (C.reltablespace = TS.oid)
            UNION
            SELECT
              TS.spcname AS tbl_space,
              schemaname,
              statio.relname AS tablename,
              NULL AS indexname,
              pg_relation_size(relid) AS size,
              heap_blks_read AS rel_blks_read,
              heap_blks_hit AS relk_blks_hit,
              heap_blks_hit+heap_blks_read AS rel_blks,
              CASE WHEN (heap_blks_hit+heap_blks_read)>0 THEN round(100*heap_blks_hit/(heap_blks_hit+heap_blks_read)) ELSE 0 END AS hit_pct
            FROM pg_catalog.pg_statio_user_tables statio
              JOIN pg_catalog.pg_class C ON (C.oid = relid)
              LEFT JOIN
                pg_catalog.pg_tablespace TS ON (C.reltablespace = TS.oid)"
  metrics:
    - tbl_space:
        usage: "LABEL"
        description: "Tablespace of this object"       
    - schemaname:
        usage: "LABEL"
        description: "Schema name of this object"       
    - tablename:
        usage: "LABEL"
        description: "If object is a table, its name. If object is an index, the table it belongs to."       
    - indexname:
        usage: "LABEL"
        description: "If object is an index, its name." 
    - size:
        usage: "GAUGE"
        description: "Size of object in bytes. Tables do not include indexes in their size."      
    - rel_blks_read:
        usage: "GAUGE"
        description: "Number of disk blocks read from this object"       
    - relk_blks_hit:
        usage: "GAUGE"
        description: "Number of buffer hits in this object"        
    - rel_blks:
        usage: "GAUGE"
        description: "Total block count for this object"
    - hit_pct:
        usage: "GAUGE"
        description: "Buffer hit percentage for this object"       
```

**Comment by wrouesnel** [comment]
The problem creeps in when the number of labels changes between samples. It's raised internally by the Prometheus client_golang code to mark generally doing something bad (because labels for a metric shouldn't change between scrapes).

At the moment my suspicion is this is triggered because "NULL" fields are actually dropped rather then returned as empty labels.

**Comment by keithf4** [comment]
That makes sense. The "indexname" column in my query is NULL for tables. Will try changing it to a default value for now. Thanks!

**Comment by keithf4** [comment]
Hmm... I tried using COALESCE and replacing the NULL column value on the other side of the UNION with a string and it's still giving the same errors.
```
ccp_statio_user_objects:
  query: "SELECT COALESCE(TS.spcname, '<<NULL>>') AS tbl_space,
              schemaname,
              statio.relname AS tablename,
              statio.indexrelname AS indexname,
              pg_relation_size(indexrelid) AS size,
              idx_blks_read AS rel_blks_read,
              idx_blks_hit AS rel_blks_hit,
              idx_blks_hit+idx_blks_read AS rel_blks,
              CASE WHEN (idx_blks_hit+idx_blks_read)>0 THEN round(100*idx_blks_hit/(idx_blks_hit+idx_blks_read)) ELSE 0 END AS hit_pct
            FROM pg_statio_user_indexes statio
              JOIN pg_class C ON (C.oid = indexrelid)
              LEFT JOIN
                pg_tablespace TS ON (C.reltablespace = TS.oid)
            UNION
            SELECT
              COALESCE(TS.spcname, '<<NULL>>') AS tbl_space,
              schemaname,
              statio.relname AS tablename,
              '<<NULL>>' AS indexname,
              pg_relation_size(relid) AS size,
              heap_blks_read AS rel_blks_read,
              heap_blks_hit AS relk_blks_hit,
              heap_blks_hit+heap_blks_read AS rel_blks,
              CASE WHEN (heap_blks_hit+heap_blks_read)>0 THEN round(100*heap_blks_hit/(heap_blks_hit+heap_blks_read)) ELSE 0 END AS hit_pct
            FROM pg_catalog.pg_statio_user_tables statio
              JOIN pg_catalog.pg_class C ON (C.oid = relid)
              LEFT JOIN
                pg_catalog.pg_tablespace TS ON (C.reltablespace = TS.oid)"
  metrics:
    - tbl_space:
        usage: "LABEL"
        description: "Tablespace of this object"       
    - schemaname:
        usage: "LABEL"
        description: "Schema name of this object"       
    - tablename:
        usage: "LABEL"
        description: "If object is a table, its name. If object is an index, the table it belongs to."       
    - indexname:
        usage: "LABEL"
        description: "If object is an index, its name." 
    - size:
        usage: "GAUGE"
        description: "Size of object in bytes. Tables do not include indexes in their size."      
    - rel_blks_read:
        usage: "GAUGE"
        description: "Number of disk blocks read from this object"       
    - relk_blks_hit:
        usage: "GAUGE"
        description: "Number of buffer hits in this object"        
    - rel_blks:
        usage: "GAUGE"
        description: "Total block count for this object"
    - hit_pct:
        usage: "GAUGE"
        description: "Buffer hit percentage for this object" 
```

**Comment by wrouesnel** [comment]
Okay so I think I've figured this out - kind of an obvious bug actually - when the exporter hits an unknown metric and there's a metric mapping in effect (which has crept into the included queries a bit more now that we've extended the functionality) then it would try to output an unknown metric with the label mapping set, but without label names being attached to the descriptor.

I'm guessing this is what's happening above - you've got an extra column in there, and so it tries to output the metric as an unknown and blows up.

I'm pushing a fix soon, and I'll do a release ASAP (it'll be 0.3.1).
